projects/driver_controls/src/x86/ads1015.c

```diff
@@ -1,28 +1,127 @@
+#include <string.h>
 #include "ads1015_def.h"
-#include "gpio_it.h"
+#include "soft_timer.h"
 
-// Initiates ads1015 by setting up registers and enabling ALRT/RDY Pin.
+#define ADS1015_CHANNEL_UPDATE_PERIOD 10
+#define ADS1015_CHANNEL_ARBITRARY_READING 0
+
+// Checks if a channel is enabled (true) or disabled (false).
+static bool prv_channel_is_enabled(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (storage == NULL || channel >= NUM_ADS1015_CHANNELS) {
+    return false;
+  }
+  return ((storage->channel_bitset & (1 << channel)) != 0);
+}
+
+// Updates the channel_bitset when a channel is enabled/disabled.
+static void prv_mark_channel_enabled(Ads1015Channel channel, bool enable, uint8_t *channel_bitset) {
+  if (channel >= NUM_ADS1015_CHANNELS || channel_bitset == NULL) {
+    return;
+  }
+  if (enable) {
+    *channel_bitset |= (1 << channel);
+  } else {
+    *channel_bitset &= ~(1 << channel);
+  }
+}
+
+// Sets the current channel of the storage.
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS || storage == NULL) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  storage->current_channel = channel;
+  return STATUS_CODE_OK;
+}
+
+// Periodically calls channels' callbacks imitating the interrupt behavior.
+static void prv_timer_callback(SoftTimerID id, void *context) {
+  if (context == NULL) {
+    return;
+  }
+  Ads1015Storage *storage = context;
+  Ads1015Channel current_channel = storage->current_channel;
+  uint8_t channel_bitset = storage->channel_bitset;
+
+  if (prv_channel_is_enabled(storage, current_channel)) {
+    storage->channel_readings[current_channel] = ADS1015_CHANNEL_ARBITRARY_READING;
+    // Runs the users callback if not NULL.
+    if (storage->channel_callback[current_channel] != NULL) {
+      storage->channel_callback[current_channel](current_channel,
+                                                 storage->callback_context[current_channel]);
+    }
+  }
+
+  prv_mark_channel_enabled(current_channel, false, &storage->pending_channel_bitset);
+  if (storage->pending_channel_bitset == ADS1015_BITSET_EMPTY) {
+    // Reset the pending bitset once gone through a cycle of channel rotation.
+    storage->pending_channel_bitset = channel_bitset;
+  }
+  current_channel = __builtin_ffs(storage->pending_channel_bitset) - 1;
+  // Update so that the ADS1015 reads from the next channel.
+  prv_set_channel(storage, current_channel);
+  soft_timer_start(ADS1015_CHANNEL_UPDATE_PERIOD, prv_timer_callback, storage, NULL);
+}
+
+// Inits the storage for ADS1015 and starts the soft timer.
 StatusCode ads1015_init(Ads1015Storage *storage, I2CPort i2c_port, Ads1015Address i2c_addr,
                         GPIOAddress *ready_pin) {
-  return status_code(STATUS_CODE_UNIMPLEMENTED);
+  if ((storage == NULL) || (ready_pin == NULL)) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  memset(storage, 0, sizeof(*storage));
```

Comment:
Nit: Personal preference that sizeof is always called with a typename so `Ads1015Storage` just to be explicit.